solver/llbsolver/solver.go

```diff
@@ -647,42 +645,79 @@ func runCacheExporters(ctx context.Context, exporters []RemoteCacheExporter, j *
-	meta := map[string][]byte{}
+func runInlineCacheExporter(ctx context.Context, e exporter.ExporterInstance, inlineExporter inlineCacheExporter, j *solver.Job, cached *result.Result[solver.CachedResult]) (*result.Result[*exptypes.InlineCacheEntry], error) {
 	if inlineExporter == nil {
 		return nil, nil
 	}
-	if err := inBuilderContext(ctx, j, "preparing layers for inline cache", j.SessionID+"-cache-inline", func(ctx context.Context, _ session.Group) error {
-		if res := cached.Ref; res != nil {
-			dtic, err := inlineCache(ctx, inlineExporter.Exporter, res, e.Config().Compression(), session.NewGroup(j.SessionID))
-			if err != nil {
-				return err
-			}
-			if dtic != nil {
-				meta[exptypes.ExporterInlineCache] = dtic
-			}
+
+	done := progress.OneOff(ctx, "preparing layers for inline cache")
+	res, err := result.ConvertResult(cached, func(res solver.CachedResult) (*exptypes.InlineCacheEntry, error) {
+		dtic, err := inlineCache(ctx, inlineExporter, res, e.Config().Compression(), session.NewGroup(j.SessionID))
+		if err != nil {
+			return nil, err
 		}
-		for k, res := range cached.Refs {
-			dtic, err := inlineCache(ctx, inlineExporter.Exporter, res, e.Config().Compression(), session.NewGroup(j.SessionID))
-			if err != nil {
-				return err
-			}
-			if dtic != nil {
-				meta[fmt.Sprintf("%s/%s", exptypes.ExporterInlineCache, k)] = dtic
+		if dtic == nil {
+			return nil, nil
+		}
+		return &exptypes.InlineCacheEntry{Data: dtic}, nil
+	})
+	return res, done(err)
+}
+
+func (s *Solver) runExporters(ctx context.Context, exporters []exporter.ExporterInstance, inlineCacheExporter inlineCacheExporter, job *solver.Job, cached *result.Result[solver.CachedResult], inp *result.Result[cache.ImmutableRef]) (exporterResponse map[string]string, descrefs map[string]exporter.DescriptorReference, err error) {
+	eg, ctx := errgroup.WithContext(ctx)
+	resps := make([]map[string]string, len(exporters))
+	descs := make([]exporter.DescriptorReference, len(exporters))
+	for i, exp := range exporters {
+		i, exp := i, exp
+		eg.Go(func() error {
+			id := fmt.Sprint(job.SessionID, "-export-", i)
+			return inBuilderContext(ctx, job, exp.Name(), id, func(ctx context.Context, _ session.Group) error {
+				inlineCache := exptypes.InlineCache(func(ctx context.Context) (*result.Result[*exptypes.InlineCacheEntry], error) {
+					return runInlineCacheExporter(ctx, exp, inlineCacheExporter, job, cached)
+				})
+
+				resps[i], descs[i], err = exp.Export(ctx, inp, inlineCache, job.SessionID)
+				if err != nil {
+					return err
+				}
+				return nil
+			})
+		})
+	}
+	if err := eg.Wait(); err != nil {
+		return nil, nil, err
+	}
+
+	// TODO: separate these out, and return multiple exporter responses to the
+	// client
+	for _, resp := range resps {
+		for k, v := range resp {
+			if exporterResponse == nil {
+				exporterResponse = make(map[string]string)
 			}
+			exporterResponse[k] = v
 		}
```

Comment:
This is the only aspect of the overall design I'm a little unsure on. I think for now, I'd be ok this, since we already have the precedent with multiple cache exporters (see the similar comment I've added there). However, long term, I think we should split out the exporter response more (while keeping [this merging behavior](https://github.com/moby/buildkit/blob/cab7db27074d275af1eeeaf12c3bd7bc8266cb88/solver/llbsolver/solver.go#L582-L591) for the deprecated field):

```protobuf
message SolveResponse {
	map<string, string> ExporterResponseDeprecated = 1;

	repeated ExporterResponse ExporterResponses = 2;
	repeated ExporterCacheResponse CacheExporterResponses = 2;
	FrontendResponse FrontendRessponse = 3;
}
```

If we take this option, I'm happy to follow-up asap with a PR to do this, but I don't think the handling for this belongs in this one, for two reasons: 1. #2760 has already been open for more than a year and 2. we should fix this for the multiple cache exporters as well, which has already merged and been released in v0.11.

Looking through the original PR, I see notice this discussed previously: https://github.com/moby/buildkit/pull/2760#issuecomment-1084350240. This is also a viable approach, where we could make sure that exporters didn't overwrite each other's keys - though this is potentially a non-backwards compatible change (depending on details).